crates/control/src/game_controller_state_filter.rs

```diff
@@ -50,59 +48,101 @@ impl GameStateFilter {
     }
 
     pub fn cycle(&mut self, context: CycleContext) -> Result<MainOutputs> {
-        let ball_detected_far_from_any_goal = ball_detected_far_from_any_goal(
+        let game_states = filter_game_states(
             *context.robot_to_field,
             context.ball_position,
             context.field_dimensions,
-            context.config.whistle_acceptance_goal_distance,
-        );
-
-        self.state = next_filtered_state(
-            self.state,
-            context.game_controller_state,
-            context.filtered_whistle.is_detected,
-            context.cycle_time.start_time,
             context.config,
-            ball_detected_far_from_any_goal,
-        );
-
-        self.opponent_state = next_filtered_state(
-            self.opponent_state,
             context.game_controller_state,
-            context.filtered_whistle.is_detected,
-            context.cycle_time.start_time,
-            context.config,
-            ball_detected_far_from_any_goal,
+            context.filtered_whistle,
+            context.cycle_time,
+            self.state,
+            self.opponent_state,
         );
+        let filtered_game_controller_state = FilteredGameControllerState {
+            game_state: game_states.filtered_game_state,
+            opponent_game_state: game_states.filtered_opponent_game_state,
+            game_phase: context.game_controller_state.game_phase,
+            kicking_team: context.game_controller_state.kicking_team,
+            penalties: context.game_controller_state.penalties,
+            remaining_number_of_messages: context
+                .game_controller_state
+                .remaining_amount_of_messages,
+            sub_state: context.game_controller_state.sub_state,
+            own_team_is_home_after_coin_toss: context
+                .game_controller_state
+                .hulks_team_is_home_after_coin_toss,
+        };
+        Ok(MainOutputs {
+            filtered_game_controller_state: Some(filtered_game_controller_state).into(),
+        })
+    }
+}
 
-        let ball_detected_far_from_kick_off_point = context
-            .ball_position
-            .map(|ball| {
-                let absolute_ball_position = *context.robot_to_field * ball.position;
-                distance(&absolute_ball_position, &Point2::origin())
-                    > context.config.distance_to_consider_ball_moved_in_kick_off
-            })
-            .unwrap_or(false);
+struct FilteredGameStates {
+    pub filtered_game_state: FilteredGameState,
+    pub filtered_opponent_game_state: FilteredGameState,
```

Comment:
```suggestion
    own: FilteredGameState,
    opponent: FilteredGameState,
```

The `pub` might not be needed. Maybe this is a leftover after removing the `pub` for the entire struct? 
Additionally, the field names, can be shortened. Following the "do not repeat yourself" rule, the type should not be repeated in the variable name.

Reply:
Done